### ISSUE TITLE
If vm is reverted to a snapshot while container is running, volume remains in attached state after revert finishes.

### ISSUE THREAD

**Issue by ashahi1** [vmware/docker-volume-vsphere#1649]
1. Took two snapshots.
2. Ran a container with volume mounted.
3. With container running, reverted to latest snaphot.
4. After revert finishes, volume remains attached, hence volume removal also fails and we have a left-over volume.

Logs have been attached at the end.

---------------------------------------

Volume name: VolSS1

```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume inspect VolSS1
[
    {
        "Driver": "vsphere:latest",
        "Labels": {},
        "Mountpoint": "/mnt/vmdk/VolSS1",
        "Name": "VolSS1",
        "Options": {},
        "Scope": "global",
        "Status": {
            "access": "read-write",
            "attach-as": "independent_persistent",
            "attached to VM": "ubuntu-VM0.1",
            "attachedVMDevice": {
                "ControllerPciSlotNumber": "160",
                "Unit": "0"
            },
            "capacity": {
                "allocated": "15MB",
                "size": "100MB"
            },
            "clone-from": "None",
            "created": "Thu Jul 27 18:27:38 2017",
            "created by VM": "ubuntu-VM0.1",
            "datastore": "sharedVmfs-0",
            "diskformat": "thin",
            "fstype": "ext4",
            "status": "attached"
        }
    }
]
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~# docker volume rm VolSS1
Error response from daemon: unable to remove volume: remove VolSS1: VolumeDriver.Remove: Failed to remove volume VolSS1, in use by VM = ubuntu-VM0.1.
root@sc-rdops-vm02-dhcp-52-237:~#

[root@sc-rdops-vm07-dhcp-192-56:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume     Datastore     VMGroup   Capacity  Used  Filesystem  Policy  Disk Format  Attached-to   Access      Attach-as               Created By    Created Date
---------  ------------  --------  --------  ----  ----------  ------  -----------  ------------  ----------  ----------------------  ------------  ------------------------
TestVolSS  sharedVmfs-0  _DEFAULT  100MB     15MB  ext4        N/A     thin         ubuntu-VM0.1  read-write  independent_persistent  ubuntu-VM0.1  Thu Jul 27 18:07:03 2017
VolSS1     sharedVmfs-0  _DEFAULT  100MB     15MB  ext4        N/A     thin         ubuntu-VM0.1  read-write  independent_persistent  ubuntu-VM0.1  Thu Jul 27 18:27:38 2017

```

Logs:
[docker-volume-vsphere.txt](https://github.com/vmware/docker-volume-vsphere/files/1180970/docker-volume-vsphere.txt)
[vmdk_ops.txt](https://github.com/vmware/docker-volume-vsphere/files/1180969/vmdk_ops.txt)

**Comment by lipingxue** [comment]
@ashahi1 The description and steps you mentioned is not clear. Please list all the steps in detail.

**Comment by govint** [comment]
@ashahi1 should also include docker ps -a  output. the container may still be running. When the VM resumes after the snapshot docker may restart the container.

**Comment by ashahi1** [comment]
Setup already had two volumes - TestVolSS@sharedVmfs-0 and VolSS1@sharedVmfs-0

@govint I did revert to previous snapshot.

Test steps:
1. Took two snapshots of the vm  - SS_1 and SS_2
2. Created a volume.
3. Ran a container and wrote data to it.
4. While container is running, reverted to latest snapshot which is SS_2.
5. After revert finished successfully, no containers are running - docker ps and docker ps -a does not show any containers.  ```docker volume inspect``` still shows volume as attached. 
6. Also volume removal fails as volume is now attached.

-----------------------

Steps and their output are as follows:

1. Created a volume SS_Volume1
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create -d vsphere:latest --name SS_Volume1
SS_Volume1
root@sc-rdops-vm02-dhcp-52-237:~# docker volume inspect SS_Volume1
[
    {
        "Driver": "vsphere:latest",
        "Labels": {},
        "Mountpoint": "/mnt/vmdk/SS_Volume1",
        "Name": "SS_Volume1",
        "Options": {},
        "Scope": "global",
        "Status": {
            "access": "read-write",
            "attach-as": "independent_persistent",
            "capacity": {
                "allocated": "13MB",
                "size": "100MB"
            },
            "clone-from": "None",
            "created": "Fri Jul 28 17:34:27 2017",
            "created by VM": "ubuntu-VM0.3",
            "datastore": "sharedVmfs-0",
            "diskformat": "thin",
            "fstype": "ext4",
            "status": "detached"
        }
    }
]
root@sc-rdops-vm02-dhcp-52-237:~#
```

2. Admin cli volume ls and docker volume ls
```
[root@sc-rdops-vm07-dhcp-214-15:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume      Datastore     VMGroup   Capacity  Used  Filesystem  Policy  Disk Format  Attached-to   Access      Attach-as               Created By    Created Date
----------  ------------  --------  --------  ----  ----------  ------  -----------  ------------  ----------  ----------------------  ------------  ------------------------
TestVolSS   sharedVmfs-0  _DEFAULT  100MB     15MB  ext4        N/A     thin         ubuntu-VM0.1  read-write  independent_persistent  ubuntu-VM0.1  Thu Jul 27 18:07:03 2017
VolSS1      sharedVmfs-0  _DEFAULT  100MB     15MB  ext4        N/A     thin         ubuntu-VM0.1  read-write  independent_persistent  ubuntu-VM0.1  Thu Jul 27 18:27:38 2017
SS_Volume1  sharedVmfs-0  _DEFAULT  100MB     13MB  ext4        N/A     thin         detached      read-write  independent_persistent  ubuntu-VM0.3  Fri Jul 28 17:34:27 2017

[root@sc-rdops-vm07-dhcp-214-15:~]


root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      SS_Volume1@sharedVmfs-0
vsphere:latest      TestVolSS@sharedVmfs-0
vsphere:latest      VolSS1@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#

```
3. Ran a container with volume mounted and wrote data to it.
```
root@sc-rdops-vm02-dhcp-52-237:~# docker run -it --volume-driver=vsphere:latest -v SS_Volume1@sharedVmfs-0:/vol1 --name ub_SS ubuntu
root@3e45e4677fbb:/#
root@3e45e4677fbb:/# echo 'Hello'>/vol1/hellos
root@3e45e4677fbb:/# cat vol1/hellos
Hello
root@3e45e4677fbb:/#
```

4. Admin cli volume ls while container is running.
```
[root@sc-rdops-vm07-dhcp-214-15:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume      Datastore     VMGroup   Capacity  Used  Filesystem  Policy  Disk Format  Attached-to   Access      Attach-as               Created By    Created Date
----------  ------------  --------  --------  ----  ----------  ------  -----------  ------------  ----------  ----------------------  ------------  ------------------------
TestVolSS   sharedVmfs-0  _DEFAULT  100MB     15MB  ext4        N/A     thin         ubuntu-VM0.1  read-write  independent_persistent  ubuntu-VM0.1  Thu Jul 27 18:07:03 2017
VolSS1      sharedVmfs-0  _DEFAULT  100MB     15MB  ext4        N/A     thin         ubuntu-VM0.1  read-write  independent_persistent  ubuntu-VM0.1  Thu Jul 27 18:27:38 2017
SS_Volume1  sharedVmfs-0  _DEFAULT  100MB     15MB  ext4        N/A     thin         ubuntu-VM0.3  read-write  independent_persistent  ubuntu-VM0.3  Fri Jul 28 17:34:27 2017

[root@sc-rdops-vm07-dhcp-214-15:~]
```

5. Reverted to the latest snapshot - SS_2

6. After the revert finished successfully, ran following commands. I do not see any container running.
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      SS_Volume1@sharedVmfs-0
vsphere:latest      TestVolSS@sharedVmfs-0
vsphere:latest      VolSS1@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~# docker volume inspect SS_Volume1@sharedVmfs-0
[
    {
        "Driver": "vsphere:latest",
        "Labels": null,
        "Mountpoint": "/mnt/vmdk/SS_Volume1@sharedVmfs-0",
        "Name": "SS_Volume1@sharedVmfs-0",
        "Options": {},
        "Scope": "global",
        "Status": {
            "access": "read-write",
            "attach-as": "independent_persistent",
            "attached to VM": "ubuntu-VM0.3",
            "attachedVMDevice": {
                "ControllerPciSlotNumber": "160",
                "Unit": "0"
            },
            "capacity": {
                "allocated": "15MB",
                "size": "100MB"
            },
            "clone-from": "None",
            "created": "Fri Jul 28 17:34:27 2017",
            "created by VM": "ubuntu-VM0.3",
            "datastore": "sharedVmfs-0",
            "diskformat": "thin",
            "fstype": "ext4",
            "status": "attached"
        }
    }
]
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~# docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
root@sc-rdops-vm02-dhcp-52-237:~# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
root@sc-rdops-vm02-dhcp-52-237:~#
```
7. Tried removing volume - failed as volume is attached
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume rm inspect SS_Volume1@sharedVmfs-0
Error response from daemon: get inspect: no such volume
Error response from daemon: unable to remove volume: remove SS_Volume1@sharedVmfs-0: VolumeDriver.Remove: Failed to remove volume SS_Volume1, in use by VM = ubuntu-VM0.3.
root@sc-rdops-vm02-dhcp-52-237:~#
```

**Comment by govint** [comment]
@ashahi1, thanks for the repro, this workflow is simply not handled in the plugin. When the revert is done, the VM goes back to a state in time (guess the snap isn't including memory either) for on-disk data. So, the VM will resume from the snapshot and docker won't even find any meta-data for the container - not even the even image if docker pulled the image after SS_2 was made.

Couple of ways to handle this is to:
1. Update the VMListener to watch for alarms and handle alarms on snapshot revert. On a snapshot revert the service can go ahead and remove the volumes from the VM config. But the catch is how does the ESX service know which volumes to detach and which volumes to keep? When a VM is reverted to a snapshot, there may be containers that may continue to run after the revert - because the snap was taken when those containers were running. Thats tricky and I'm not sure if this approach is even doable.
1. The other approach is to update the Get() request to the ESX service to accept a "nil" volume name and return meta-data for all those volumes that are attached to the VM.

1. The plugin on start up makes the call to the ESX service to get all attached volumes and uses that to decide which volumes to keep or which ones to unmount.
1. This is more robust and fits well with the refcount recovery scheme.
1. The refcount recovery is assuming that the plugin is restarted without the VM getting restarted. Using the approach proposed here it ensures we are able to handle all kind of restarts of the plugin and the VM.

**Comment by govint** [comment]
@ashahi1 I tried the below tests,

Test1 - Run the container after taking the snaps

1. Make two snaps of a VM
vim-cmd vmsvc/snapshot.get 1
Get Snapshot:
|-ROOT
--Snapshot Name        : master-1
--Snapshot Id        : 13
--Snapshot Desciption  :
--Snapshot Created On  : 7/31/2017 10:23:29
--Snapshot State       : powered off
--|-CHILD
----Snapshot Name        : master-2
----Snapshot Id        : 14
----Snapshot Desciption  :
----Snapshot Created On  : 7/31/2017 10:23:32
----Snapshot State       : powered off

2. Now, start a container thats attaching a VDVS volume
# docker run -it --volume-driver=vsphere -v snapvol1:/vol1 ubuntu
root@fbff02fa99e6:/#

3.  Check the disks attached to the VM
 grep vmdk *vmx
scsi0:0.fileName = "master-VM0.0-000002.vmdk"
scsi0:1.fileName = "/vmfs/volumes/59474fd4-d6871ae4-c86b-02001e5a2f6d/vmdk/a-000002.vmdk"
scsi1:0.fileName = "/vmfs/volumes/59474fd4-d6871ae4-c86b-02001e5a2f6d/dockvols/11111111-1111-1111-1111-111111111111/snapvol1.vmdk" <--------------

4. Revert to the second snap
 govc snapshot.revert -vm master-VM0.0 -s=false master-2

4. Check disks attached to the VM
grep vmdk *vmx
scsi0:0.fileName = "master-VM0.0-000003.vmdk"
scsi0:1.fileName = "/vmfs/volumes/59474fd4-d6871ae4-c86b-02001e5a2f6d/vmdk/a-000003.vmdk"

So, the revert has correctly removed the VMDK that was attached _after_ the snap was taken.

5. But the volume is still marked as attached because VDVS service hasn't detached the volume by itself.
# docker volume inspect snapvol1
[
    {
        "Driver": "vsphere:latest",
        "Labels": {},
        "Mountpoint": "/mnt/vmdk/snapvol1",
        "Name": "snapvol1",
        "Options": {},
        "Scope": "global",
        "Status": {
            "access": "read-write",
            "attach-as": "independent_persistent",
            "attached to VM": "master-VM0.0",
            "attachedVMDevice": {
                "ControllerPciSlotNumber": "160",
                "Unit": "0"
            },
            "capacity": {
                "allocated": "15MB",
                "size": "100MB"
            },
            "clone-from": "None",
            "created": "Mon Jul 31 08:41:41 2017",
            "created by VM": "master-VM0.0",
            "datastore": "sharedVmfs-1",
            "diskformat": "thin",
            "fstype": "ext4",
            "status": "attached"
        }
    }
]

Test-2: Run the container before taking a snap

1. Make a snap
vim-cmd vmsvc/snapshot.get 1
Get Snapshot:
|-ROOT
--Snapshot Name        : master-1
--Snapshot Id        : 15
--Snapshot Desciption  :
--Snapshot Created On  : 7/31/2017 10:34:27
--Snapshot State       : powered off

2. Run a container with VDVS volume
docker run -it --volume-driver=vsphere -v snapvol1:/vol1 ubuntu
root@d32e124ae7f7:/#

3. Take another snap
vim-cmd vmsvc/snapshot.get 1
Get Snapshot:
|-ROOT
--Snapshot Name        : master-1
--Snapshot Id        : 15
--Snapshot Desciption  :
--Snapshot Created On  : 7/31/2017 10:34:27
--Snapshot State       : powered off
--|-CHILD
----Snapshot Name        : master-2
----Snapshot Id        : 16
----Snapshot Desciption  :
----Snapshot Created On  : 7/31/2017 10:36:27
----Snapshot State       : powered off

4. Revert to the first snap _master-1_, which was taken before container was run

5. Check the vmdks attached to the VM

 grep vmdk *vmx
scsi0:0.fileName = "master-VM0.0-000003.vmdk"
scsi0:1.fileName = "/vmfs/volumes/59474fd4-d6871ae4-c86b-02001e5a2f6d/vmdk/a-000003.vmdk"

6. Check volume data, still shows attached to the VM,
docker volume inspect snapvol1
[
    {
        "Driver": "vsphere",
        "Labels": null,
        "Mountpoint": "/mnt/vmdk/snapvol1",
        "Name": "snapvol1",
        "Options": {},
        "Scope": "global",
        "Status": {
            "access": "read-write",
            "attach-as": "independent_persistent",
            "attached to VM": "master-VM0.0",
            "attachedVMDevice": {
                "ControllerPciSlotNumber": "160",
                "Unit": "0"
            },
            "capacity": {
                "allocated": "15MB",
                "size": "100MB"
            },
            "clone-from": "None",
            "created": "Mon Jul 31 08:41:41 2017",
            "created by VM": "master-VM0.0",
            "datastore": "sharedVmfs-1",
            "diskformat": "thin",
            "fstype": "ext4",
            "status": "attached"
        }
    }
]

Test-3: Run container before taking a snap and revert to the newest snap

1. Make a snap
vim-cmd vmsvc/snapshot.get 1
Get Snapshot:
|-ROOT
--Snapshot Name        : master-1
--Snapshot Id        : 17
--Snapshot Desciption  :
--Snapshot Created On  : 7/31/2017 10:49:4
--Snapshot State       : powered off

2. Run container and check attached disks for VM
grep vmdk *vmx
scsi0:0.fileName = "master-VM0.0-000001.vmdk"
scsi0:1.fileName = "/vmfs/volumes/59474fd4-d6871ae4-c86b-02001e5a2f6d/vmdk/a-000001.vmdk"
scsi1:0.fileName = "/vmfs/volumes/59474fd4-d6871ae4-c86b-02001e5a2f6d/dockvols/11111111-1111-1111-1111-111111111111/snapvol1.vmdk"

3. Make one more snap
vim-cmd vmsvc/snapshot.get 1
Get Snapshot:
|-ROOT
--Snapshot Name        : master-1
--Snapshot Id        : 17
--Snapshot Desciption  :
--Snapshot Created On  : 7/31/2017 10:49:4
--Snapshot State       : powered off
--|-CHILD
----Snapshot Name        : master-2
----Snapshot Id        : 18
----Snapshot Desciption  :
----Snapshot Created On  : 7/31/2017 10:50:2
----Snapshot State       : powered off

4. Revert to snap taken in (3)
govc snapshot.tree -vm master-VM0.0
master-1
  master-2
  .

govc snapshot.revert -vm master-VM0.0 -s=false master-2

5. Check disks attached to VM
grep vmdk *vmx
scsi0:0.fileName = "master-VM0.0-000003.vmdk"
scsi0:1.fileName = "/vmfs/volumes/59474fd4-d6871ae4-c86b-02001e5a2f6d/vmdk/a-000003.vmdk"
scsi1:0.fileName = "/vmfs/volumes/59474fd4-d6871ae4-c86b-02001e5a2f6d/dockvols/11111111-1111-1111-1111-111111111111/snapvol1.vmdk"

6. Check VM state after snapshot revert

vim-cmd vmsvc/power.getstate 1
Retrieved runtime info
Powered off

VM is powered off but volume is still attached, the snapshot revert will not remove the disk because the snapshot was taken after the container was started and running. But the VM is powered off and still VMLIstener hasn't detached the disk. VMListener doesn't get a notification of this power off as part of snapshot revert.

**Comment by govint** [comment]
From the above update there are two issues:

1. Issue 1 - When a snapshot revert operation removes a VDVS volume from the VM config, the VMDK KV file has the still marked as attached. This is _not an issue_, as the next attach will anyway _override_ the state in the KV and will allow the volume to be attached to the next requestor.

2. Issue 2 - When a container is running and a snap is taken and then user reverts back to this snap

1. The VM is powered off as a result of the revert op then
1. The power-off event isn't being received by the VMListener
1. The vmdk is left attached to the VM.

Couple of ways to handle this is to:
Update the VMListener to watch for alarms and handle alarms on snapshot revert. On a snapshot revert the service can go ahead and remove the volumes from the VM config. But its tricky if the VM listener thread races with containers in the docker host that are attaching VDVS managed volumes (post-revert).

The other approach is to update the Get() request to the ESX service to accept a "nil" volume name and return meta-data for all those volumes that are attached to the VM. Or the List() call perhaps.
1.	The plugin on start up makes the call to the ESX service to get all attached volumes and uses that to decide which volumes to keep or which ones to unmount.
2.	This is more robust and fits well with the refcount recovery scheme.
3.	The refcount recovery is assuming that the plugin is restarted without the VM getting restarted. Using the approach proposed here it ensures we are able to handle all kind of restarts of the plugin and the VM.


Note:
1. Not related, but once a container is running, a snapshot that saves VM memory state can't be made if independent disks (VDVS volumes default attach mode) are attached to the VM.
1. It should be possible to include VM mem. state in the snapshot if container disks are attached in persistent mode. Which means containers can be running across revert ops and the VM listener can't necessarily detach volumes on a snapshot revert event.

**Comment by govint** [comment]
Effort - 0.5 (if only documentation) else as noted in #1675

**Comment by govint** [comment]
This issue is being fixed via PRs #1689 and #1710.

**Comment by ashahi1** [comment]
Fixed as part of PR #1689
